content/blog/2025-03-20-new-hash-table.md

```diff
@@ -0,0 +1,202 @@
++++
+title= "A new hash table"
+date= 2025-03-20 00:00:00
+description= "Designing a state-of-the art hash table implementation"
+authors= [ "zuiderkwast", "SoftlyRaining"]
++++
+
+Valkey is essentially a giant hash table attached to the network. A hash table
+is the data structure that maps keys to values. When optimizing for latency, CPU
+and memory usage, it's natural to look at the hash table internals. By replacing
+the hash table with a different implementation, we have managed to reduce the
+memory usage by roughly 20 bytes per key-value pair and improve the latency and
+CPU usage by rougly 10% for instances without I/O threading.
+
+Results
+-------
+
+Memory usage for keys of length N and value of length M bytes. TBD.
+
+| Version    | Memory usage per key   |
+|------------|------------------------|
+| Valkey 7.2 | ? bytes                |
+| Valkey 8.0 | ? bytes                |
+| Valkey 8.1 | ? bytes                |
+
+The benchmarks below were run using a key size of N and a value size of M bytes, without pipelining.
+
+| Command                 | Valkey 7.2 | Valkey 8.0 | Valkey 8.1 |
+|-------------------------|------------|------------|------------|
+| SET                     | Xµs, Y QPS | ?          | ?          |
+| GET                     | Xµs, Y QPS | ?          | ?          |
+| ...                     | ...        | ?          | ?          |
+
+The benchmark was run on an xxxx using yyyy, without I/O threads.
+
+Background
+----------
+
+The slowest operation when looking up a key-value pair is by far reading from
+the main RAM memory. A key point when optimizing a hash table is therefore to
+make sure we have as few memory accesses as possible. Ideally, the memory
+reading is already in the CPU cache, which is much faster memory that belong to
+the CPU.
```

Comment:
I think this background is OK, but I think it would be beneficial to show the previous implementation, and then explain how we iterated on it to the new one. Then it becomes easier to show the memory accesses that are getting removed. You already show the previous design later on, so we can probably move it earlier.

A good general framing is, show the current state (the dict), highlight the gaps (lots of memory accesses and allocations), introduce a new concept that solves that (cache friendly), then discuss the solution. Most of that content is here, just needs a little bit of moving around.